### ISSUE TITLE
match failed: Value too large for defined data type

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-sched#1129]
In preparation for some possible scaling tests, I was running a test Flux instance with 48 brokers per node across 64 nodes on corona. When trying to run a job with one task per rank, Fluxion failed with the following error:

```
$ flux run -N3072 -n3072 hostname
Dec 27 08:59:43.540844 sched-fluxion-resource.err[0]: match_multi_request_cb: match failed due to match error (id=4541860806656): Value too large for defined data type
0.112s: job.exception type=alloc severity=0 alloc denied due to type="match error"
```

FYI - 32 nodes with 48 brokers per node does work.

**Comment by grondo** [comment]
Here's a standalone reproducer. The problem seems to require a larger number of cores per node. That is, I was able to run this test with 48 cores per node up to 10000 nodes. However, with 112 cores per node, the max `TEST_SIZE` that can be used here is 813:
```sh
#!/bin/bash

TEST_SIZE=${TEST_SIZE:-2000}

flux module remove -f sched-fluxion-qmanager
flux module remove -f sched-fluxion-resource

flux module remove resource
flux config load <<EOF
[sched-fluxion-qmanager]
queue-policy = "easy"
#[sched-fluxion-resource]
#match-format = "$MATCH_FORMAT"

[resource]
noverify = true
norestrict = true

[[resource.config]]
hosts = "test[1-${TEST_SIZE}]"
cores = "0-112"
gpus = "0-8"
EOF

flux module load resource monitor-force-up
flux module load sched-fluxion-resource
flux module load sched-fluxion-qmanager
flux queue start --all --quiet
flux resource list
flux resource status

flux run -vvv -N${TEST_SIZE} -n${TEST_SIZE} \
	--setattr=exec.test.run_duration=1ms \
	true

flux jobs -a
```

```console
$ TEST_SIZE=813 flux start ./test.sh 
     STATE NNODES   NCORES    NGPUS NODELIST
      free    813    91869     7317 test[1-813]
 allocated      0        0        0 
      down      0        0        0 
     STATE UP NNODES NODELIST
     avail  ✔    813 test[1-813]
jobid: f3m7fSF1
0.000s: job.submit {"userid":6885,"urgency":16,"flags":0,"version":1}
0.013s: job.validate
0.024s: job.depend
0.024s: job.priority {"priority":16}
0.095s: job.alloc
0.095s: job.prolog-start {"description":"cray-pals-port-distributor"}
0.096s: job.cray_port_distribution {"ports":[11999,11998],"random_integer":5801541299176206611}
0.096s: job.prolog-finish {"description":"cray-pals-port-distributor","status":0}
0.098s: job.start
0.099s: job.finish {"status":0}
0.096s: exec.init
0.098s: exec.starting
0.099s: exec.complete {"status":0}
0.099s: exec.done
flux-job: No job output found
       JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
    f3m7fSF1 grondo   true       CD    813    813   0.004s test[1-813]
```

```console
$ TEST_SIZE=814 flux start ./test.sh 
     STATE NNODES   NCORES    NGPUS NODELIST
      free    814    91982     7326 test[1-814]
 allocated      0        0        0 
      down      0        0        0 
     STATE UP NNODES NODELIST
     avail  ✔    814 test[1-814]
jobid: f3nNJq3u
0.000s: job.submit {"userid":6885,"urgency":16,"flags":0,"version":1}
0.013s: job.validate
0.024s: job.depend
0.024s: job.priority {"priority":16}
Dec 27 13:57:50.461517 sched-fluxion-resource.err[0]: match_request_cb: match failed due to match error (id=105914564608): Value too large for defined data type
0.049s: job.exception type=alloc severity=0 alloc denied due to type="match error"
0.049s: job.clean
       JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
    f3nNJq3u grondo   true        F    814    814        - 
```

**Comment by grondo** [comment]
Tagging @milroy

**Comment by milroy** [comment]
I can reproduce the problem on corona with the script above. I'm investigating this now.

**Comment by milroy** [comment]
I can also reproduce the error on quartz (where I have an active bank account and don't have to run on a login node). Interestingly, I can run the reproducer successfully with a test size of 10,000 (presumably beyond) if I specify 8 (`gpus = "0-7"`) rather than 9 GPUs/per node:

```bash
bash-4.4$ TEST_SIZE=10000 flux start ./test.sh 
     STATE NNODES   NCORES    NGPUS NODELIST
      free  10000  1130000    80000 test[1-10000]
 allocated      0        0        0 
      down      0        0        0 
     STATE UP NNODES NODELIST
     avail  ✔  10000 test[1-10000]
jobid: fPNmJKeF
0.000s: job.submit {"userid":31186,"urgency":16,"flags":0,"version":1}
0.012s: job.validate
0.023s: job.depend
0.023s: job.priority {"priority":16}
1.218s: job.alloc
1.218s: job.prolog-start {"description":"cray-pals-port-distributor"}
1.218s: job.cray_port_distribution {"ports":[11999,11998],"random_integer":6861328851445906924}
1.218s: job.prolog-finish {"description":"cray-pals-port-distributor","status":0}
1.230s: job.start
1.231s: job.finish {"status":0}
1.229s: exec.init
1.230s: exec.starting
1.231s: exec.complete {"status":0}
1.231s: exec.done
flux-job: No job output found
       JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
    fPNmJKeF milroy1  true       CD  10000  10000   0.013s test[1-10000]
```

I suspect this is a GPU ID mapping issue.

**Comment by milroy** [comment]
Even more interesting: I can successfully run a test size of 814 with 10 GPUs/node.

**Comment by milroy** [comment]
There's a problem with overflow and modular arithmetic in `calc_factor`. @grondo can you try this change and see if it fixes the behavior (it did in my testing):

```diff
diff --git a/resource/policies/dfu_match_multilevel_id_impl.hpp b/resource/policies/dfu_match_multilevel_id_impl.hpp
index 8b629940..c7716323 100644
--- a/resource/policies/dfu_match_multilevel_id_impl.hpp
+++ b/resource/policies/dfu_match_multilevel_id_impl.hpp
@@ -55,7 +55,7 @@ int64_t multilevel_id_t<FOLD>::score_factor_t::calc_factor (
         return -1;
     }
     mul = add * m_multiply_by;
-    tie =  break_tie % m_multiply_by - 1;
+    tie = abs (break_tie % static_cast<int64_t> (m_multiply_by) - 1);
 
     if (mul > (std::numeric_limits<int64_t>::max () - tie)) {
         errno = EOVERFLOW;
```

**Comment by grondo** [comment]
This should have been closed by #1132. Feel free to reopen if that's incorrect.

**Comment by garlick** [comment]
I'm noticing these in elcapi's logs where we have flux-sched=0.33.0:
```
Apr 01 13:49:50 elcap1 flux[149853]: sched-fluxion-resource.err[0]: match_request_cb: match failed due to match error (id=334578204402515968): Value too large for defined data type
```

```
$ flux job info 334578204402515968 eventlog
{"timestamp":1712004590.0767584,"name":"submit","context":{"userid":57277,"urgency":16,"flags":0,"version":1}}
{"timestamp":1712004590.0956175,"name":"validate"}
{"timestamp":1712004590.1097062,"name":"depend"}
{"timestamp":1712004590.1097784,"name":"priority","context":{"priority":16}}
{"timestamp":1712004590.4934578,"name":"exception","context":{"type":"alloc","severity":0,"note":"alloc denied due to type=\"match error\""}}
{"timestamp":1712004590.4935486,"name":"clean"}
```
